File: rtms_api/python_manual_start_stop_rtms/rtms.py
import os
import json
import asyncio
import websockets
import requests
import hmac
import hashlib
from flask import Flask, request, jsonify
from dotenv import load_dotenv

# Load environment variables from .env
load_dotenv()

app = Flask(__name__)

# Step 1: Webhook Receiver - Listen for meeting events
@app.route("/webhook", methods=['POST'])
def webhook():
    data = request.get_json()
    event = data.get('event')
    payload = data.get('payload')
    
    print(f'Webhook received: {event}')
    print(f'Payload: {json.dumps(payload, indent=2)}')

    # Step 2a: Listen to meeting started event
    if event == 'meeting.started':
        print('Meeting started, initiating RTMS...')
        meeting_object = payload.get('object', {})
        meeting_id = meeting_object.get('id')
        meeting_uuid = meeting_object.get('uuid')
        
        try:
            # Step 2b: Get access token
            access_token = generate_access_token()
            
            # Step 2c: Make API call to start RTMS
            start_rtms(meeting_id, access_token)
            
            print(f'RTMS started for meeting {meeting_id}')
            
            # Schedule automatic RTMS stop after 60 seconds
            schedule_rtms_stop(meeting_id, access_token)
        except Exception as error:
            print(f'Error starting RTMS: {error}')

    # Step 3: RTMS started event
    if event == 'meeting.rtms_started':
        print('Starting RTMS connection...')
        meeting_uuid = payload.get('meeting_uuid')
        rtms_stream_id = payload.get('rtms_stream_id')
        server_urls = payload.get('server_urls')
        asyncio.run(connect_to_signaling_websocket(meeting_uuid, rtms_stream_id, server_urls))

    # When meeting RTMS stops, log the stop event
    if event == 'meeting.rtms_stopped':
        meeting_uuid = payload.get('meeting_uuid')
        print(f'Meeting {meeting_uuid} stopped')
        # Open WebSocket connections will close naturally

    return jsonify({'status': 'success'}), 200

# Step 4: Generate Signature for authentication handshake
def generate_signature(meeting_uuid, stream_id):
    client_id = os.getenv('ZOOM_CLIENT_ID')
    client_secret = os.getenv('ZOOM_CLIENT_SECRET')
    
    if not client_id or not client_secret:
        raise Exception('ZOOM_CLIENT_ID and ZOOM_CLIENT_SECRET must be set in environment variables')
    
    message = f"{client_id},{meeting_uuid},{stream_id}"
    signature = hmac.new(
        client_secret.encode('utf-8'),
        message.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    return signature

# Helper function: Generate access token from environment variable
def generate_access_token():
    access_token = os.getenv('access_token')
    if not access_token:
        raise Exception('access_token not found in environment variables')
    print('Using access token from environment variables')
    return access_token

# Helper function: Manually start RTMS using Zoom API
def start_rtms(meeting_id, access_token):
    try:
        url = f"https://api.zoom.us/v2/live_meetings/{meeting_id}/rtms_app/status"
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {access_token}'
        }
        data = {
            'action': 'start',
            'settings': {
                'client_id': os.getenv('ZOOM_CLIENT_ID')
            }
        }
        
        response = requests.patch(url, json=data, headers=headers)
        
        # Better error handling
        if response.status_code != 200:
            print(f'API Error: Status {response.status_code}')
            print(f'Response: {response.text}')
            raise Exception(f'API call failed with status {response.status_code}')
        
        try:
            result = response.json()
            print(f'RTMS start response: {result}')
            return result
        except json.JSONDecodeError:
            print(f'Invalid JSON response: {response.text}')
            raise Exception('Invalid JSON response from API')
            
    except requests.exceptions.RequestException as error:
        print(f'Error starting RTMS via API: {error}')
        raise error

# Step 5: WebSocket 1 - Connect to Signaling WebSocket
async def connect_to_signaling_websocket(meeting_uuid, stream_id, server_urls):
    print(f'Connecting to signaling WebSocket: {server_urls}')
    
    try:
        # Disable SSL certificate verification for development
        import ssl
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE
        
        async with websockets.connect(server_urls, ssl=ssl_context) as signaling_ws:
            print('Signaling WebSocket opened')
            
            # Send handshake
            handshake_message = {
                'msg_type': 1,  # HANDSHAKE_REQUEST
                'meeting_uuid': meeting_uuid,
                'rtms_stream_id': stream_id,
                'signature': generate_signature(meeting_uuid, stream_id)
            }
            await signaling_ws.send(json.dumps(handshake_message))
            
            # Handle incoming signaling messages
            async for message in signaling_ws:
                msg = json.loads(message)
                print(f'Signaling message: {msg}')

                # If handshake is successful, proceed to media connection
                if msg.get('msg_type') == 2 and msg.get('status_code') == 0:
                    media_url = msg['media_server']['server_urls']['transcript']
                    await connect_to_media_websocket(media_url, meeting_uuid, stream_id, signaling_ws)

                # If keep-alive request is received, respond with ACK
                if msg.get('msg_type') == 12:
                    ack_message = {
                        'msg_type': 13,
                        'timestamp': msg.get('timestamp')
                    }
                    await signaling_ws.send(json.dumps(ack_message))
                    
    except Exception as error:
        print(f'Signaling WebSocket error: {error}')
    finally:
        print('Signaling WebSocket closed')

# Step 6: WebSocket 2 - Connect to Media WebSocket
async def connect_to_media_websocket(media_url, meeting_uuid, stream_id, signaling_socket):
    print(f'Connecting to media WebSocket: {media_url}')
    
    try:
        # Disable SSL certificate verification for development
        import ssl
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE
        
        async with websockets.connect(media_url, ssl=ssl_context) as media_ws:
            print('Media WebSocket opened')
            
            # Send media handshake
            media_handshake = {
                'msg_type': 3,  # MEDIA_HANDSHAKE_REQUEST
                'protocol_version': 1,
                'sequence': 0,
                'meeting_uuid': meeting_uuid,
                'rtms_stream_id': stream_id,
                'signature': generate_signature(meeting_uuid, stream_id),
                'media_type': 8  # Request transcript stream
            }
            await media_ws.send(json.dumps(media_handshake))
            
            # Handle incoming media messages
            async for message in media_ws:
                msg = json.loads(message)
                print(f'Media message received: {msg}')

                # Respond to keep-alive request from media server
                if msg.get('msg_type') == 12:
                    print('Responding to media keep-alive')
                    ack_message = {
                        'msg_type': 13,
                        'timestamp': msg.get('timestamp')
                    }
                    await media_ws.send(json.dumps(ack_message))

                # If media handshake is successful, notify signaling server that client is ready
                if msg.get('msg_type') == 4 and msg.get('status_code') == 0:
                    print('Media handshake successful, sending CLIENT_READY_ACK')
                    ready_message = {
                        'msg_type': 7,
                        'rtms_stream_id': stream_id
                    }
                    await signaling_socket.send(json.dumps(ready_message))

                # Log incoming transcript data
                if msg.get('msg_type') == 5:
                    print(f'Transcript: {msg}')
                    
    except Exception as error:
        print(f'Media WebSocket error: {error}')
    finally:
        print('Media WebSocket closed')

# Step 7: Stop RTMS using Zoom API
def stop_rtms(meeting_id, access_token):
    try:
        url = f"https://api.zoom.us/v2/live_meetings/{meeting_id}/rtms_app/status"
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {access_token}'
        }
        data = {
            'action': 'stop',
            'settings': {
                'client_id': os.getenv('ZOOM_CLIENT_ID')
            }
        }
        
        response = requests.patch(url, json=data, headers=headers)
        response.raise_for_status()
        
        print(f'RTMS stop response: {response.json()}')
        return response.json()
    except requests.exceptions.RequestException as error:
        print(f'Error stopping RTMS via API: {error}')
        raise error

# Helper function: Schedule RTMS stop after 10 seconds
def schedule_rtms_stop(meeting_id, access_token):
    print(f'Scheduling RTMS stop for meeting {meeting_id} in 10 seconds...')
    
    def stop_after_delay():
        import threading
        import time
        
        def delayed_stop():
            time.sleep(10)  # 10 seconds delay
            try:
                print(f'Stopping RTMS for meeting {meeting_id}...')
                stop_rtms(meeting_id, access_token)
                print(f'RTMS stopped successfully for meeting {meeting_id}')
            except Exception as error:
                print(f'Failed to stop RTMS for meeting {meeting_id}: {error}')
        
        thread = threading.Thread(target=delayed_stop)
        thread.daemon = True
        thread.start()
    
    stop_after_delay()

# Step 8: Start Flask server on port 3000
if __name__ == '__main__':
    print('Server running on port 3000')
    app.run(host='0.0.0.0', port=3000, debug=True)

File: rtms_api/manual_start_stop_using_rtms_api/rtms.js
import express from 'express';
import WebSocket from 'ws';
import crypto from 'crypto';
import dotenv from 'dotenv';
import axios from 'axios';

// Load environment variables from .env
dotenv.config();

const app = express();
app.use(express.json()); // Parse incoming JSON payloads

// Step 1: Webhook Receiver - Listen for meeting events
app.post("/webhook", async (req, res) => {
  const { event, payload } = req.body;
  console.log('Webhook received:', event);
  console.log('Payload:', JSON.stringify(payload, null, 2));

  // Step 2a: Listen to meeting started event
  if (event === 'meeting.started') {
    console.log('Meeting started, initiating RTMS...');
    const { object } = payload;
    const meetingId = object.id;
    const meetingUuid = object.uuid;
    
    try {
      // Step 2b: Get access token
      const accessToken = generateAccessToken();
      
      // Step 2c: Make API call to start RTMS
      await startRTMS(meetingId, accessToken);
      
      console.log(`RTMS started for meeting ${meetingId}`);
      
      // Schedule automatic RTMS stop after 60 seconds
      scheduleRTMSStop(meetingId, accessToken);
    } catch (error) {
      console.error('Error starting RTMS:', error);
    }
  }

  // Step 3: RTMS started event
  if (event === 'meeting.rtms_started') {
    console.log('Starting RTMS connection...');
    const { meeting_uuid, rtms_stream_id, server_urls } = payload;
    connectToSignalingWebSocket(meeting_uuid, rtms_stream_id, server_urls);
  }

  // When meeting RTMS stops, log the stop event
  if (event === 'meeting.rtms_stopped') {
    const { meeting_uuid } = payload;
    console.log(`Meeting ${meeting_uuid} stopped`);
    // Open WebSocket connections will close naturally
  }

  res.sendStatus(200); // Acknowledge webhook receipt
});

// Step 4: Generate Signature for authentication handshake
function generateSignature(meetingUuid, streamId) {
  const message = `${process.env.ZOOM_CLIENT_ID},${meetingUuid},${streamId}`;
  return crypto.createHmac('sha256', process.env.ZOOM_CLIENT_SECRET)
    .update(message).digest('hex');
}

// Helper function: Generate access token from environment variable
// Your logic to generate accesstoken will be here. Please visit https://developers.zoom.us/docs/integrations/oauth/ to know more.
function generateAccessToken() {
  const accessToken = process.env.access_token;
  if (!accessToken) {
    throw new Error('access_token not found in environment variables');
  }
  console.log('Using access token from environment variables');
  return accessToken;
}

// Helper function: Manually start RTMS using Zoom API
async function startRTMS(meetingId, accessToken) {
  try {
    const response = await axios.patch(
      `https://api.zoom.us/v2/live_meetings/${meetingId}/rtms_app/status`,
      {
        action: 'start',
        settings: {
          client_id: process.env.ZOOM_CLIENT_ID
        }
      },
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${accessToken}`
        }
      }
    );
    
    console.log('RTMS start response:', response.data);
    return response.data;
  } catch (error) {
    console.error('Error starting RTMS via API:', error.response?.data || error.message);
    throw error;
  }
}

// Step 5: WebSocket 1 - Connect to Signaling WebSocket
function connectToSignalingWebSocket(meetingUuid, streamId, serverUrls) {
  console.log(`Connecting to signaling WebSocket: ${serverUrls}`);
  const signalingWs = new WebSocket(serverUrls);

  // Once signaling WebSocket is open, send handshake
  signalingWs.on('open', () => {
    console.log('Signaling WebSocket opened');
    signalingWs.send(JSON.stringify({
      msg_type: 1, // HANDSHAKE_REQUEST
      meeting_uuid: meetingUuid,
      rtms_stream_id: streamId,
      signature: generateSignature(meetingUuid, streamId)
    }));
  });

  // Handle incoming signaling messages
  signalingWs.on('message', (data) => {
    const msg = JSON.parse(data);
    console.log('Signaling message:', msg);

    // If handshake is successful, proceed to media connection
    if (msg.msg_type === 2 && msg.status_code === 0) {
      const mediaUrl = msg.media_server.server_urls.transcript;
      connectToMediaWebSocket(mediaUrl, meetingUuid, streamId, signalingWs);
    }

    // If keep-alive request is received, respond with ACK
    if (msg.msg_type === 12) {
      signalingWs.send(JSON.stringify({
        msg_type: 13,
        timestamp: msg.timestamp
      }));
    }
  });

  // Log signaling errors
  signalingWs.on('error', (error) => {
    console.error('Signaling WebSocket error:', error);
  });

  // Log signaling WebSocket closure
  signalingWs.on('close', () => {
    console.log('Signaling WebSocket closed');
  });
}

// Step 6: WebSocket 2 - Connect to Media WebSocket
function connectToMediaWebSocket(mediaUrl, meetingUuid, streamId, signalingSocket) {
  console.log(`Connecting to media WebSocket: ${mediaUrl}`);
  const mediaWs = new WebSocket(mediaUrl);

  // Once media WebSocket is open, send media handshake
  mediaWs.on('open', () => {
    console.log('Media WebSocket opened');
    mediaWs.send(JSON.stringify({
      msg_type: 3, // MEDIA_HANDSHAKE_REQUEST
      protocol_version: 1,
      sequence: 0,
      meeting_uuid: meetingUuid,
      rtms_stream_id: streamId,
      signature: generateSignature(meetingUuid, streamId),
      media_type: 8 // Request transcript stream
    }));
  });

  // Handle incoming media messages
  mediaWs.on('message', (data) => {
    const msg = JSON.parse(data);
    console.log('Media message received:', msg);

    // Respond to keep-alive request from media server
    if (msg.msg_type === 12) {
      console.log('Responding to media keep-alive');
      mediaWs.send(JSON.stringify({
        msg_type: 13,
        timestamp: msg.timestamp
      }));
    }

    // If media handshake is successful, notify signaling server that client is ready
    if (msg.msg_type === 4 && msg.status_code === 0) {
      console.log('Media handshake successful, sending CLIENT_READY_ACK');
      signalingSocket.send(JSON.stringify({
        msg_type: 7,
        rtms_stream_id: streamId
      }));
    }

    // Log incoming transcript data
    if (msg.msg_type === 5) {
      console.log('Transcript:', msg);
    }
  });

  // Log media errors
  mediaWs.on('error', (error) => {
    console.error('Media WebSocket error:', error);
  });

  // Log media WebSocket closure
  mediaWs.on('close', () => {
    console.log('Media WebSocket closed');
  });
}


// Step 7: Stop RTMS using Zoom API
async function stopRTMS(meetingId, accessToken) {
  try {
    const response = await axios.patch(
      `https://api.zoom.us/v2/live_meetings/${meetingId}/rtms_app/status`,
      {
        action: 'stop',
        settings: {
          client_id: process.env.ZOOM_CLIENT_ID
        }
      },
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${accessToken}`
        }
      }
    );
    
    console.log('RTMS stop response:', response.data);
    return response.data;
  } catch (error) {
    console.error('Error stopping RTMS via API:', error.response?.data || error.message);
    throw error;
  }
}

// Helper function: Schedule RTMS stop after 10 seconds 
// Your logic to stop RTMS will go here 
function scheduleRTMSStop(meetingId, accessToken) {
  console.log(`Scheduling RTMS stop for meeting ${meetingId} in 60 seconds...`);
  
  setTimeout(async () => {
    try {
      console.log(`Stopping RTMS for meeting ${meetingId}...`);
      await stopRTMS(meetingId, accessToken);
      console.log(`RTMS stopped successfully for meeting ${meetingId}`);
    } catch (error) {
      console.error(`Failed to stop RTMS for meeting ${meetingId}:`, error);
    }
  }, 10000); // 10 seconds
}

// Step 7: Start Express server on port 3000
app.listen(3000, () => console.log('Server running on port 3000'));

